### PR TITLE
Add onboarding prompt when home feed too slow in web UI

### PR DESCRIPTION
**Pull request by Gargron**
If the most recent posts on the home feed are from fewer than 7 different accounts, or the average gap between posts is more than 20 minutes, or the most recent post is from more than an hour ago, show a dismissable banner that points to trends and follow recommendations.

Screenshot 2023-06-22 at 17-28-18 Localhost

I've also taken the opportunity to fix some of the very awkwardly worded dismissable banners in the app. I've found that "social web" is a great descriptor for the fediverse, so now:

|Before|After|
|--|--|
|These posts from this and other servers in the decentralized network are gaining traction on this server right now.|These are posts from across the social web that are gaining traction today. Newer posts with more boosts and favourites are ranked higher.|
|These hashtags are gaining traction among people on this and other servers of the decentralized network right now.|These are hashtags that are gaining traction on the social web today. Hashtags that are used by more different people are ranked higher.|
|These news stories are being talked about by people on this and other servers of the decentralized network right now.|These are news stories being shared the most on the social web today. Newer news stories posted by more different people are ranked higher.|

The new descriptions even touch on how the results are sorted.
___

Fixes MAS-155

### CODE REVIEW

File: app/javascript/mastodon/features/home_timeline/components/explore_prompt.jsx [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import { FormattedMessage } from 'react-intl';
+
+import { Link } from 'react-router-dom';
+
+import DismissableBanner from 'mastodon/components/dismissable_banner';
+
+export const ExplorePrompt = () => (
+  <DismissableBanner id='home.explore_prompt'>
+    <p><FormattedMessage id='home.explore_prompt' defaultMessage="Your home feed is all the people you follow and the posts they share. It becomes more useful when you follow more people. It looks like it's not very lively yet, so how about:" /></p>
```
Comment (ThisIsMissEm) [inline app/javascript/mastodon/features/home_timeline/components/explore_prompt.jsx:9]: Home feed also includes posts from hashtags you follow, doesn't it?

Comment (Gargron) [reply app/javascript/mastodon/features/home_timeline/components/explore_prompt.jsx:9]: Yeah, but that level of detail is probably out of place?

Comment (ThisIsMissEm) [reply app/javascript/mastodon/features/home_timeline/components/explore_prompt.jsx:9]: I think people are commonly confused by that, so if we're telling people something, it probably would be good to be accurate, especially as posts from followed hashtags don't get labelled as such when in the home timeline.

**Comment by shleeable** [comment]
This is a nice touch. I NEVER find myself visiting the trending view, and that's a real shame.... so anything that reminds people to visit that is a positive to me.

**Comment by jazmichaelking** [comment]
Some possible copy edits to reduce word count, improve readability:

Posts on the social web that are popular today. Newer posts with more boosts and favourites are ranked higher.

Hashtags on the social web that are popular today. Hashtags used by more people are ranked higher.

News stories shared the most on the social web today. Newer stories posted by more people are ranked higher.

Deeper edits option:

Today's popular posts. Newer posts with more boosts and favourites are ranked higher.

Today's popular hashtags. Hashtags used by more people are ranked higher.

Today's popular news. Newer stories posted by more people are ranked higher.

File: app/javascript/mastodon/features/home_timeline/index.jsx [hunk 2]
```diff
@@ -28,12 +32,43 @@ const messages = defineMessages({
...
+    const bucket = Math.floor(cur.getTime() / precision);
+
+    if (acc[bucket]) {
+      acc[bucket] += 1;
+    } else {
+      acc[bucket] = 1;
+    }
+
+    return acc;
+  }, {})).reduce((p, c, i) => p + (c - p) / (i + 1), 0);
```
Comment (ClearlyClaire) [inline app/javascript/mastodon/features/home_timeline/index.jsx:52]: This bit is honestly hard to follow and I'm not quite sure what it's meant to compute.

I think the following does the same thing while being significantly easier to read.
```suggestion
  const tally = Object.values(statuses.map(status => new Date(status.get('created_at'))).reduce((tally, created_at) => {
    const bucket = Math.floor(created_at.getTime() / precision);
    tally[bucket] = (tally[bucket] || 0) + 1;
    return tally;
  }, {}));
  const values = Object.values(tally);
  const frequency = values.length ? values.reduce((total, value) => total + value) / values.length : 0;
```

But is that really what you *want* to compute? This is the average number of posts per bucket of 20 minutes, but not counting empty buckets.

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]

File: app/javascript/mastodon/features/explore/links.jsx [hunk 3]
```diff
@@ -35,7 +35,7 @@ class Links extends PureComponent {
 
     const banner = (
       <DismissableBanner id='explore/links'>
-        <FormattedMessage id='dismissable_banner.explore_links' defaultMessage='These news stories are being talked about by people on this and other servers of the decentralized network right now.' />
+        <FormattedMessage id='dismissable_banner.explore_links' defaultMessage='These are news stories being shared the most on the social web today. Newer news stories posted by more different people are ranked higher.' />
```
Comment (ClearlyClaire) [inline app/javascript/mastodon/features/explore/links.jsx:38]: I'm slightly afraid this may be misleading people into thinking this is what's trending on the *whole* fediverse when it's “only” the posts that are known to your own server (which is going to be a huge difference for small-ish servers), but this was already a concern with the previous wording, and it's difficult striking a good balance between concise and understandable and precise enough.

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]

File: app/javascript/mastodon/features/home_timeline/index.jsx [hunk 4]
```diff
@@ -28,12 +31,36 @@ const messages = defineMessages({
...
+    gap: averageGap,
+    last: newest,
+  };
+});
+
+const homeTooSlow = createSelector(getHomeFeedSpeed, speed =>
+  speed.unique < 7 // If there are fewer than 7 different accounts visible
+  || speed.gap > (20 * 60) // If the average gap between posts is more than 20 minutes
+  || (Date.now() - speed.last) > (1000 * 60 * 60) // If the most recent post is from over an hour ago
+);
```
Comment (ClearlyClaire) [inline app/javascript/mastodon/features/home_timeline/index.jsx:55]: Looking at these metrics…
- I believe the 20 last posts being made by at least 7 different people is very likely to trigger even for people with a very active feed
- this will likely trigger in the morning for most people who do not follow people from many timezones
- more generally, the whole “less than a post every 20 minutes is too slow” approach is a really worrying mindset, though I understand the point of the feature, and the wording of the banner is not bad

Comment (Gargron) [reply app/javascript/mastodon/features/home_timeline/index.jsx:55]: I tried basing this on my feed, though of course it depends on time of day. I don't follow that many people (413).

Comment (Gargron) [reply app/javascript/mastodon/features/home_timeline/index.jsx]: Changed to 5 unique accounts and 30 min gap.

Comment (ClearlyClaire) [reply app/javascript/mastodon/features/home_timeline/index.jsx]: I think the 5 unique accounts thing needs to go. Otherwise, “It's looking pretty quiet right now” will pop up when a limited set of followed people are having a lively conversation, which does not make a lot of sense.

**Comment by parnikkapore** [comment]
Just chiming in that the buttons on the onboarding prompt are truncated in the Advanced / multicolumn UI on mastodon.social; unfortunately, I've dismissed it already, so cannot provide a screenshot.

**Comment by ocdtrekkie** [comment]
I noticed it appears briefly every time I load the page and disappears as my timeline loads. Advanced view as well.

**Comment by Oreolek** [comment]
I think this will be depressing if you're not speaking English. Imagine someone coming in and the UI is all translated but the trends are all in English and even after finding people to talk to (probably all on the same instance) the website still thinks you're not trying hard enough because the community is too small to count.
